static/app/views/dashboardsV2/layoutUtils.tsx

```diff
@@ -75,3 +85,74 @@ export function getDashboardLayout(widgets: Widget[]): Layout[] {
       i: constructGridItemKey(widget),
     }));
 }
+
+export function pickDefinedStoreKeys(layout: Layout): Partial<Layout> {
+  return pickBy(layout, (value, key) => defined(value) && STORE_KEYS.includes(key));
+}
+
+export function getWidgetHeight(displayType: DisplayType): number {
+  return displayType === DisplayType.BIG_NUMBER ? 1 : 2;
+}
+
+/**
+ * Creates an array from layouts where each column stores how deep it is.
+ */
+export function calculateColumnDepths(layouts: Layout[]): number[] {
+  const depths = Array(NUM_DESKTOP_COLS).fill(0);
+
+  // For each layout's x, record the max depth
+  layouts
+    .filter(({i}) => i !== ADD_WIDGET_BUTTON_DRAG_ID)
+    .forEach(({x, w, y, h}) => {
+      // Adjust the column depths for each column the widget takes up
+      for (let col = x; col < x + w; col++) {
+        depths[col] = Math.max(y + h, depths[col]);
+      }
+    });
+
+  return depths;
+}
+
+/**
+ * Find the next place to place a widget and also returns the next
+ * input when this operation needs to be called multiple times.
+ *
+ * @param columnDepths A profile of how deep the widgets in a column extend.
+ * @param height The desired height of the next widget we want to place.
+ * @returns An {x, y} positioning for the next available spot, as well as the
+ * next columnDepths array if this position were used.
+ */
+export function getNextAvailablePosition(
+  columnDepths: number[],
+  height: number
+): [Position, number[]] {
```

Comment:
Can we use named tuples here? https://www.typescriptlang.org/play?ts=4.0.2#example/named-tuples